src/_P044_P1WifiGateway.ino

```diff
@@ -5,127 +5,223 @@
+#define P044_DATAGRAM_START_CHAR           '/'
+#define P044_DATAGRAM_END_CHAR             '!'
+#define P044_DATAGRAM_MAX_SIZE             1024
 #define P044_NETBUF_SIZE                   128
-#define P044_DISABLED                      0
-#define P044_WAITING                       1
-#define P044_READING                       2
-#define P044_CHECKSUM                      3
-#define P044_DONE                          4
 
+#define P044_WIFI_SERVER_PORT     ExtraTaskSettings.TaskDevicePluginConfigLong[0]
+#define P044_BAUDRATE             ExtraTaskSettings.TaskDevicePluginConfigLong[1]
+#define P044_RX_WAIT              PCONFIG(0)
+#define P044_SERIAL_CONFIG        PCONFIG(1)
+#define P044_RESET_TARGET_PIN     CONFIG_PIN1
+ 
 
+struct P044_Task : public PluginTaskData_base {
 
-struct P044_data_struct : public PluginTaskData_base {
+  enum ParserState : byte {
+    P044_WAITING,
+    P044_READING,
+    P044_CHECKSUM
+  };
 
-  P044_data_struct() {
+  P044_Task() {
     clearBuffer();
   }
 
-  ~P044_data_struct() {
+  ~P044_Task() {
     stopServer();
   }
 
+  inline static bool serverActive(WiFiServer *server) {
+#if defined(ESP8266)
+    return nullptr != server && server->status() != CLOSED;
+#elif defined(ESP32)
+    return nullptr != server && *server;
+#endif
+  }
+
   void startServer(unsigned int portnumber) {
+    if (gatewayPort == portnumber && serverActive(P1GatewayServer)) {
+      // server is already listening on this port
+      return;
+    }
     stopServer();
+    gatewayPort = portnumber;
     P1GatewayServer = new WiFiServer(portnumber);
-    if (nullptr != P1GatewayServer) {
+    if (nullptr != P1GatewayServer && WiFi.isConnected()) {
       P1GatewayServer->begin();
-      addLog(LOG_LEVEL_DEBUG, String(F("P1   : WiFi server started at port ")) + portnumber);
+      if(serverActive(P1GatewayServer)) {
+        addLog(LOG_LEVEL_INFO, String(F("P1   : WiFi server started at port ")) + portnumber);
+      } else {
+        addLog(LOG_LEVEL_ERROR, String(F("P1   : WiFi server start failed at port ")) +
+            portnumber + String(F(", retrying...")));
+      }
+    }
+  }
+
+  void checkServer() {
+    if (!serverActive(P1GatewayServer) && WiFi.isConnected()) {
+      P1GatewayServer->close();
+      P1GatewayServer->begin();
+      if(serverActive(P1GatewayServer)) {
+        addLog(LOG_LEVEL_INFO, F("P1   : WiFi server started"));
+      }
     }
   }
 
   void stopServer() {
-    clearBuffer();
     if (nullptr != P1GatewayServer) {
+      if (P1GatewayClient) P1GatewayClient.stop();
       P1GatewayServer->close();
-      addLog(LOG_LEVEL_DEBUG, F("P1   : WiFi server closed"));
+      addLog(LOG_LEVEL_INFO, F("P1   : WiFi server closed"));
       delete P1GatewayServer;
       P1GatewayServer = nullptr;
     }
   }
 
-  void clearBuffer() {
-    serial_buffer = "";
-    serial_buffer.reserve(P044_BUFFER_SIZE);
-    bytes_read = 0;
+  bool hasClientConnected() {
+    if (P1GatewayServer->hasClient())
```

Comment:
No check for `nullptr`

Reply:
Already checked for when obtaining the pointer to the task, thus precondition here.

Reply:
Sanity check added